lib/value/src/kind/accessor.rs

```diff
@@ -0,0 +1,543 @@
+                        let map = BTreeMap::from([(index.into(), self)]);
+                        Collection::from(map)
+                    } else {
+                        Collection::any()
+                    };
+
+                    self = Self::array(collection);
+                }
+            }
+        }
+
+        self
+    }
+
+    /// Remove, and return the `Kind` at the given `path`.
+    ///
+    /// For arrays, indices are shifted back if any element before the last is removed.
+    ///
+    /// If the `kind` is a non-collection type, or the path points to a non-existing location in
+    /// a collection, this method returns `None`.
+    ///
+    /// Negative indexing always returns `None`, as the type system doesn't know the exact size of
+    /// an array, and thus cannot count backward.
+    ///
+    /// # Panics
+    ///
+    /// `path` must not point to root. This is because it's ambiguous whether to return the
+    /// root-level `object` or `array`, if `Kind` has both defined.
+    ///
+    /// Use `into_object` or `into_array` if you need the root-level object or array.
+    pub fn remove_at_path(&mut self, path: &Lookup<'_>) -> Option<Self> {
+        // Cannot remove using root-path.
+        if path.is_root() {
+            panic!("cannot remove root path");
+        }
+
+        let mut kind = self;
+        let mut iter = path.iter().peekable();
+
+        while let Some(segment) = iter.next() {
+            let last = iter.peek().is_none();
+
+            kind = match segment {
+                // Remove and return the final field.
+                Segment::Field(field) if last => {
+                    return kind
+                        .object
+                        .as_mut()
+                        .and_then(|collection| collection.known_mut().remove(&(field.into())))
+                }
+
+                // Try finding the field in the existing object.
+                Segment::Field(field) => kind
+                    .object
+                    .as_mut()
+                    .and_then(|collection| collection.known_mut().get_mut(&(field.into())))?,
+
+                // Remove and return the final matching field.
+                Segment::Coalesce(fields) if last => {
+                    return kind.object.as_mut().and_then(|collection| {
+                        fields
+                            .iter()
+                            .find_map(|field| collection.known_mut().remove(&(field.into())))
+                    })
+                }
+
+                // Try finding one of the fields in the existing object.
+                Segment::Coalesce(fields) => kind.object.as_mut().and_then(|collection| {
+                    let field = fields
+                        .iter()
+                        .find(|field| collection.known().contains_key(&((*field).into())))?;
+
+                    collection.known_mut().get_mut(&(field.into()))
```

Comment:
I think this should recurse down all matching fields in the coalesce to update all matching paths.

Reply:
Similar to above, I don't think it should. It should stop at the first matched field, similar to how querying a coalesced path stops at querying the first matched field.

Reply:
Given this typedef, `{thing: { thang: int } | null, thung: { thang: int } }`.

The typedef for that could represent both objects:

`{ thing: { thang: 32 }, thung: { thang: 43 } }`
`{ thung: { thang: 43 } }`

Calling `del(x, .(thing | thung).thang)` would result in:

`{ thing: {}, thung: { thang: 43 } }`
`{ thung: {} }`

Thereby the typedef for the object should represent the fact that `thang` can be removed from both branches.

If you call `remove_at_path(.(thing | thung).thang)`, currently the result would be: `{ thung: { thang: float }`, implying that `thang` can never be removed from `thung`.

Reply:
Ah yes, you are totally correct. Ugh, these coalesced paths are such a mess to work with. I'm curious if we actually handled this correctly before. Regardless, it should work as expected, so I'll look into this.

Depending on the complexity, I might do it in a follow-up PR so that I'm not blocked on moving the schema work forward for now. It _might_ be the case that I can no longer take a `&mut self` but have to take the `Kind` by value and return a cloned variant, which would be unfortunate, but the price of getting this right. Although it might still be possible to keep the mutable reference, I'll have to see how much the borrow-checker complains (implementing these functions using `&mut self` was quite a bit of work to get right, the old implementation used cloning to work around that).

Reply:
I wonder if we could just remove all type defs at the first coalesce and say anything goes after that?

Reply:
That would be a "correct, but non-complete" way to tackle this. I think I'm for this approach to start, we can refine it if we really see a need to do so.

Reply:
I've instead opted to reject coalesced path deletion (in this library) for now.